_posts/2017-06-15-grpc-load-balancing.md

```diff
@@ -0,0 +1,190 @@
+
+In L7 (application level) load balancing, the LB terminates and parses the HTTP/2 protocol. It creates a new HTTP/2 connection to the backend of choice. It then forwards the HTTP/2 streams received from the client to the backend(s) of choice. With HTTP/2 LB can distribute the streams from one client among multiple backends. LB can inspect the request and assign backend based on request contents. As an example, a session cookie sent as part of HTTP header can be used to associate with a specific backend, so all requests for that session are served by the same backend.
+
+#### L3/L4 (Transport) vs L7 (Application)
+
+<table>
+  <tr>
+    <td>
+    Use case
+    </td>
+    <td>
+    Recommendation
+    </td>
+  </tr>
+  <tr>
+    <td>RPC load varies a lot among connections</td>
+    <td>Use Application level LB</td>
+  </tr>
+  <tr>
+    <td>Storage or compute affinity is important</td>
+    <td>Use Application level LB and use cookies or similar for routing requests to correct backend</td>
+  </tr>
+  <tr>
+    <td>Minimizing resource utilization in proxy is more important than features</td>
+    <td>Use L3/L4 LB</td>
+  </tr>
+  <tr>
+    <td>Latency is paramount</td>
+    <td>Use L3/L4 LB</td>
+  </tr>
+</table>
+
+
+### Client side LB options - Thick client vs Look-aside
+
+*Note: look-aside is also known as one-arm in some literature.*
+
+#### Thick client
+
+A thick client approach means the load balancing smarts are implemented in the client. The client is responsible for keeping track of available servers, their workload, and the algorithms used for choosing servers. The client typically integrates libraries that communicate with other infrastructures - Service Discovery, Name Resolution, Quota management, etc.
+
+#### Lookaside Load Balancing
+
+A lookaside load balancer is another approach where the smarts are implemented in a different LB server. Clients query the lookaside LB and the LB responds with best server(s) to use. The heavy lifting of keeping server state and implementation of LB algorithm is consolidated in the lookaside LB. Note that client might choose to implement simple algorithms on top of the sophisticated ones implemented in the LB. GRPC defines a protocol for communication between client and LB. See this [doc](https://github.com/grpc/grpc/blob/master/doc/load-balancing.md) for details.
+
+The picture below visualizes the concept - Client gets at least one address from Lookaside LB (#1), client uses this address to make a RPC (#2), and server sends load report to the LB (#3). The Lookaside LB communicates with other infrastructure such as Name resolution, Service Discovery etc (#4).
```

Comment:
The picture below illustrates this approach. The client gets at least one address from the lookaside LB (#1). Then the client uses this address to send an RPC (#2), and the server sends its load report to the LB (#3). The lookaside LB communicates with other infrastructure such as name resolution, service discovery, and so on (#4).